### ISSUE TITLE
The Pre-trained model from `asr-transformer-transformerlm-librispeech` does not contain all information

### ISSUE THREAD

**Issue by csukuangfj** [speechbrain/speechbrain#931]
The pre-trained model in
https://huggingface.co/speechbrain/asr-transformer-transformerlm-librispeech
does not contain the state information for `speechbrain.processing.features.InputNormalization`.

See
https://github.com/speechbrain/speechbrain/blob/8070883dbf33313ab8c143e9725fb3043ba1fdf6/recipes/LibriSpeech/ASR/transformer/hparams/transformer.yaml#L230-L232

The consequence is that during inference time the WER differs if you change batch
size as the mean and std are computed using data within a batch.

**Comment by 30stomercury** [comment]
Hi @csukuangfj, I also spotted this issue. The default args: epoch, and update_until_epoch should be set properly for normalization.

**Comment by csukuangfj** [comment]
> The default args: epoch, and update_until_epoch should be set properly for normalization.

How would you recommend setting them?

I think the global mean and std should be saved into the checkpoint and during the test time, it just loads
those statistics from the checkpoint and use them directly, without depending on the test data.

---

By the way, how are the published results obtained?  Are they decoded along the training stage or are they obtained
by using the published pre-trained model?

Have the collaborators ever tried to reproduce the WERs by using the pre-trained model?

**Comment by mravanelli** [comment]
There are few more things to make the computation pipeline fully batch-independent:
1. Fix the InputNorm (and update statistics only during training). Note that is has only a very minor effect because when you load the training statistics, you also load the count (self.count) and thus the weight given to the current mean is very low (see https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/processing/features.py#L1085). However, this should be fixed. 
2. We should switch from padding_model=reflect, to padding_mode=zeros in convolutions ad STFT computation. This is one of the condition to have the same output for a sentence processed alone or within a batch.

**Comment by mravanelli** [comment]
As for the results, you can find all the logs, training, performance here:
https://drive.google.com/drive/folders/1ZVuROxR711Xib2MsJbcPla4PWqbK1Ddw?usp=sharing

The models that you find on HuggigFace are the same that you find here. The performance should be the exactly the same as the one you see in the logs after fixing the batch-dependency thing.

**Comment by csukuangfj** [comment]
> As for the results, you can find all the logs, training, performance here:
> https://drive.google.com/drive/folders/1ZVuROxR711Xib2MsJbcPla4PWqbK1Ddw?usp=sharing
> 
> The models that you find on HuggigFace are the same that you find here. The performance should be the exactly the same as the one you see in the logs after fixing the batch-dependency thing.
> 
> Also, note that there is a possibly related discussion going on here #924

Ah, I see. The results listed in the readme are obtained during training from checkpoints, NOT.from.the.provided.pre-trained model.  This may explain why the published WER cannot.be.reproduced by using pre-trained models because
checkpoints contain more information than the pre-trained models.

**Comment by csukuangfj** [comment]
I would suggest also listing the WER computed from the pre-trained model since most of us (if not all)
try to reproduce the result using only the pre-trained models. 

It can save our lives.

If someone finds that he gets a different WER with the pre-trained model from the published result,
he may think there may exist bugs in his code and will spend lots of time digging into the code and asking around.
If the WER from the pre-trained model is available from this repo, it can save us lots of time.

**Comment by mravanelli** [comment]
I have created this [PR](https://github.com/speechbrain/speechbrain/pull/933) with the goal of fixing step by step all the batch-dependent computations. Any help is welcome @csukuangfj!

**Comment by danpovey** [comment]
> There are few more things to make the computation pipeline fully batch-independent:
> 
> 1. Fix the InputNorm (and update statistics only during training). Note that is has only a very minor effect because when you load the training statistics, you also load the count (self.count) and thus the weight given to the current mean is very low (see https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/processing/features.py#L1085). However, this should be fixed.
> 2. We should switch from padding_model=reflect, to padding_mode=zeros in convolutions ad STFT computation. This is one of the condition to have the same output for a sentence processed alone or within a batch.
> 3. The convolution done on the top of a padded signal has a tail effect that we should somehow neutralize. One idea is to mask the extra non-zero steps added at the end of the convolution at the end of the [LengthsCapableSequential](https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/nnet/containers.py#L151)
> 4. Make sure the isn't anything depending of the signal length in the beamsearch part as well.

RE the InputNorm: it looks to me like, when it says
` elif epoch < self.update_until_epoch:`
it can be just changed to:
` elif epoch < self.update_until_epoch and self.training:`
which should fix the stats during inference when the mode is "global".

Also it looks to me like this code:
```
                self.glob_mean.detach()
                self.glob_std.detach()
```
is a bug; it should be: `self.glob_mean = self.glob_mean.detach()` and so on.  Right now, this code is probably a big memory leak.

**Comment by danpovey** [comment]
Also, in InputNormalization, the way it stores speaker statistics, I'm not sure if that's considered totally normal in terms of testing protocol, if you end up remembering stats from speakers in training who might recur in test time.
For the global stats, rather than saving/loading them it would be more normal I think to just register them as buffers, e.g. `self.register_buffer('glob_mean', ... )` (but you'd need to know the dimension ahead of time, I think).
This will also cause ddp to sync the copy the stats of job 0 to all other jobs (not that this should matter much, for
the global stats).

**Comment by mravanelli** [comment]
Hi @danpovey ,
thank you for the comments. The `InputNormalization` part needs a refactor (e.g, we want to vectorize more). @30stomercury  is already working on that.

1. As for the training modality yes, there is an easy way to fix that.
2. Thank you for noticing the detaching part. As you mentioned, the code is not actually leaking because we have other safeguards (e.g., `x = (x - self.glob_mean.data) / (self.glob_std.data)`). However, this is something to fix.
3. As for the speaker statistics, they aren't really used in the current recipes and we can even remove them. The idea was to allow users that have speaker information at test time to normalize the mean and std of the features based on the speaker identity (this leads to minor benefits according to my experience).
4. Registering a buffer could be something to consider as well. It will make the implementation more similar to the batch_norm one (that needs to save running averages).

Thank you again for your help!

**Comment by Adel-Moumen** [comment]
Hello, 

What should we do about this issue? Is there any ongoing work on this? CC @mravanelli 

Thanks!
